internal/eventstream/client.go

```diff
@@ -0,0 +1,336 @@
+		default: // faulty
+			eventSeverity = event.SeverityErr
+		}
+	}
+
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeState
+	ev.Severity = eventSeverity
+	ev.Message = result.Output
+
+	return ev, nil
+}
+
+// buildAcknowledgementEvent from the given fields.
+func (client *Client) buildAcknowledgementEvent(host, service, author, comment string) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeAcknowledgement
+	ev.Username = author
+	ev.Message = comment
+
+	return ev, nil
+}
+
+// eventDispatcher receives generated event.Events to be either buffered or directly delivered to the CallbackFn.
+//
+// When the Client is in the replay phase, events from the Event Stream API will be cached until the replay phase has
+// finished, while replayed events will be delivered directly.
+func (client *Client) eventDispatcher() {
+	var replayBuffer []*event.Event
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing event dispatcher as its context is done", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayTrigger:
+			for _, ev := range replayBuffer {
+				client.CallbackFn(ev)
+			}
+			client.Logger.Infow("Finished replay phase, returning to normal operation", zap.Int("cached events", len(replayBuffer)))
+			client.replayPhase.Store(false)
+			replayBuffer = []*event.Event{}
+
+		case ev := <-client.eventDispatcherEventStream:
+			if client.replayPhase.Load() {
+				replayBuffer = append(replayBuffer, ev)
+			} else {
+				client.CallbackFn(ev)
+			}
+
+		case ev := <-client.eventDispatcherReplay:
+			if !client.replayPhase.Load() {
+				client.Logger.Errorw("Dispatcher received replay event during normal operation", zap.Stringer("event", ev))
+			}
+			client.CallbackFn(ev)
+		}
+	}
+}
+
+// enterReplayPhase enters the replay phase for the initial sync and after reconnections.
+//
+// This method starts multiple goroutines. First, some workers to query the Icinga 2 Objects API will be launched. When
+// all of those have finished, the replayTrigger will be used to indicate that the buffered Events should be replayed.
+func (client *Client) enterReplayPhase() {
+	client.Logger.Info("Entering replay phase to replay stored events first")
+	if !client.replayPhase.CompareAndSwap(false, true) {
+		client.Logger.Error("The Event Stream Client is already in the replay phase")
+		return
+	}
+
+	queryFns := []func(string){client.checkMissedAcknowledgements, client.checkMissedStateChanges}
+	objTypes := []string{"host", "service"}
+
+	var replayWg sync.WaitGroup
+	replayWg.Add(len(queryFns) * len(objTypes))
+
+	for _, fn := range queryFns {
+		for _, objType := range objTypes {
+			go func(fn func(string), objType string) {
+				fn(objType)
+				replayWg.Done()
+			}(fn, objType)
+		}
+	}
+
+	go func() {
+		startTime := time.Now()
+		replayWg.Wait()
```

Comment:
You could also use [`errgroup`](https://pkg.go.dev/golang.org/x/sync/errgroup) here. With that, you don't have to do that WaitGroup tracking manually and as it's used in Icinga DB (and Icinga Go Library in the future), it won't even be a new dependency.